### ISSUE TITLE
device_tag model is not as good as expect

### ISSUE THREAD

**Issue by xinsuinizhuan** [josephjaspers/blackcat_tensors#33]
mnist_test_recurrent example 500 epochs， run it in different model:
template<class System= BC::device_tag>：
training time:2628.69
Average loss: 0.0552672

template<class System=BC::host_tag>
training time: 2704.01
Average loss: 0.0552672

the BC::host_tag is faster 76ms than BC::device_tag model，my cuda is  nvida 1060,.

**Comment by josephjaspers** [comment]
Which implementation? 
The LSTM is slower due to making copies. (And calling cudaMalloc). I am working on fixing this currently though. But I can start working on that now if you need more performance.

**Comment by josephjaspers** [comment]
Just tested to make sure..

The FeedForward mnistexample takes about 
(cpu) 16.54 seconds
(gpu)  5.4 seconds

cpu inteli5 
gpu gtx 980m 

Hopefully after I fix the LSTM model you should be able to see substantially faster results. 
(I just started a new job so I apologize for the slow progress, once I get my schedule sorted  I should be able to be more productive again).

**Comment by xinsuinizhuan** [comment]
You are so great, i think you should focus on you new job，then make your schedule sorted and work on this. I expected the lstm model GPU good performace and the attention-lstm model, the lstm model predict precision is not very accurate for my forecast, the attension-lstm is improved model,it is said the  attension-lstm is better than lstm.

**Comment by josephjaspers** [comment]
I am learning about the attention-lstm. I will write it once I understand it!

**Comment by xinsuinizhuan** [comment]
Test performance by mnist_test_recurrnet:
template<class System=BC::host_tag>
training time: 2070.37
Average loss: 0.0552672

template<class System = BC::device_tag>
training time: 2092.14
Average loss: 0.0552672

a little faster than befor, but the device_tag is slower than the host_tag, why？

**Comment by xinsuinizhuan** [comment]
I set epochs = 128, then train, by by mnist_test_recurrnetm，the BC::nn::MAPE Average loss: 0.005487, it is better then an set epochs = 500，Average loss: 0.0552672。Why?

**Comment by xinsuinizhuan** [comment]
when i load my .txt data and not use the as_derived() and utility_l::HostToDevice, it is ok?

**Comment by josephjaspers** [comment]
It is okay, it might just be slower if you are passing data to the gpu.

**Comment by josephjaspers** [comment]
> I set epochs = 128, then train, by by mnist_test_recurrnetm，the BC::nn::MAPE Average loss: 0.005487, it is better then an set epochs = 500，Average loss: 0.0552672。Why?

This occurs generally when the learning-rate is too high. It is recommended after a  few epochs to decrease the learning rate. 

This problem occurs even in more commonly used libraries 
https://github.com/keras-team/keras/issues/1971
